### ISSUE TITLE
Uploaded files are copied, not moved, from temp file to destination

### ISSUE THREAD

**Issue by Revertron** [owncloud/core#22939]
### Steps to reproduce
1. Upload big file
2. Let the upload be aborted by timeout
3. Look into 'upload_tmp_dir' directory and into owncloud directory - temporary php file will be in first one and partial file copy of it will be in target owncloud dir.

### Expected behavior
The uploaded file must be quickly moved, not slowly copied to target dir and then deleted from temp dir.

### Actual behavior
The temporary uploaded file from 'upload_tmp_dir' is (slowly) copied to target owncloud directory and can be stopped by execution timeout. But if 'upload_tmp_dir' and owncloud data dir are on the same drive it must be moved quickly, not copied.

### Server configuration
**Operating system**:
Debian 8.3 x32
**Web server:**
nginx 1.6.2
**Database:**
mysql
**PHP version:**
5.6.17-0+deb8u1 
**ownCloud version:** (see ownCloud admin page)
8.2.2 (stable)
**Updated from an older ownCloud or fresh install:**
Updated from 8.0 3-4 times
**Where did you install ownCloud from:**
From zip file (sources)
**Signing status (ownCloud 9.0 and above):**

```
Login as admin user into your ownCloud and access 
http://example.com/index.php/settings/integrity/failed 
paste the results here.
```
Redirects to start page (files app)
**List of activated apps:**

```
Enabled:
  - activity: 2.1.3
  - audios: 1.2.5
  - bookmarks: 0.7
  - calendar: 0.8.2
  - contacts: 0.5.0.0
  - direct_menu: 0.6
  - documents: 0.11.0
  - encryption: 1.1.0
  - files: 1.2.0
  - files_external: 0.3.0
  - files_locking:
  - files_pdfviewer: 0.7
  - files_sharing: 0.7.0
  - files_texteditor: 2.0
  - files_trashbin: 0.7.0
  - files_versions: 1.1.0
  - files_videoviewer: 0.1.3
  - firstrunwizard: 1.1
  - gallery: 14.2.0
  - news: 7.1.2
  - provisioning_api: 0.3.0
  - templateeditor: 0.1
  - transmission: 0.2
  - updater: 0.6
Disabled:
  - external
  - notifications
  - user_external
  - user_ldap
  - user_webdavauth
```

**The content of config/config.php:**

```
{
    "system": {
        "instanceid": "oc4p0q4g0czk",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "cloud.jamm.me"
        ],
        "datadirectory": "\/mnt\/usb2\/OwnCloud\/data",
        "overwrite.cli.url": "https:\/\/cloud.jamm.me",
        "dbtype": "mysql",
        "version": "8.2.2.2",
        "dbname": "owncloud",
        "dbhost": "127.0.0.1",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "forcessl": true,
        "forceSSLforSubdomains": true,
        "mail_from_address": "cloud",
        "mail_smtpmode": "php",
        "mail_domain": "jamm.me",
        "loglevel": 2,
        "theme": "reverse_one",
        "remember_login_cookie_lifetime": 1296000,
        "session_lifetime": 259200,
        "maintenance": false,
        "logfile": "\/var\/www\/owncloud\/owncloud_rev.log",
        "appstore.experimental.enabled": true,
        "debug": true,
        "trashbin_retention_obligation": "auto",
        "memcache.local": "\\OC\\Memcache\\APCu"
    }
}
```

**Are you using external storage, if yes which one:** local/smb/sftp/...
No
**Are you using encryption:** yes/no
No
**Are you using an external user-backend, if yes which one:** LDAP/ActiveDirectory/Webdav/...
No

### Client configuration
**Browser:**
Firefox 45.0
**Operating system:**
Windows 7 SP1 x86_64
### Logs
There are no relative logs.

**Comment by PVince81** [comment]
Are you uploading through the web UI or webdav ?

Please check your nginx settings. From what I heard there is a setting that tells nginx to first store a file to temp storage (even with Webdav) and only then pass it to the PHP process.

See https://github.com/owncloud/core/issues/9832#issuecomment-166577967

**Comment by Revertron** [comment]
@PVince81 I'm uploading through web UI.
The main problem is in PHP, it gets the file, but it is NOT MOVING, BUT COPYING file to final destination. It takes long time for large files (say 2Gb and slow HDD on USB2.0).
Does the owncloud use `move_uploaded_file()` function in the first place? Or makes a copy and then deletes temporary file?

**Comment by PVince81** [comment]
Ok, the web UI upload currently uses POST. With PHP, this means always going to temp storage.

In 9.1 we'll look into making this a PUT which should go directly to the final storage: https://github.com/owncloud/core/pull/21237

**Comment by Revertron** [comment]
What is the answer for my questions? Who makes `cp` instead of `mv`? PHP itself or owncloud has some workaround?

**Comment by PVince81** [comment]
When uploading in the web UI, PHP does a POST and stores a temporary file. Only after it finished storing this file, it will give control to ownCloud. ownCloud then needs to read that file and stream its contents to the final file. It is not possible to move the file directly to the final destination with a simple `mv` because of many cases like encryption, external storage, etc.

**Comment by PVince81** [comment]
There is currently no workaround for this for web UI upload.
In 9.1 the web UI upload will be changed to use PUT instead of POST (the usage of POST was mostly due to IE compatibility in the past), in this PR: https://github.com/owncloud/core/pull/21237

Does that answer your question ?

**Comment by Revertron** [comment]
Yes, thanks.
But you could've used simple `move_uploaded_file()` if there is no encryption. May be that would be more robust?

**Comment by MorrisJobke** [comment]
> But you could've used simple move_uploaded_file() if there is no encryption. May be that would be more robust?

Sadly it's not that simple. We need to respect all possible setups.

We just notice that when a user never logs in via the web ui the chunks of failed uploads sometimes stay a bit longer on disk than they should. To address this we have a separate ticket: https://github.com/owncloud/core/issues/14646

I will close this ticket here.
